### DISCUSSION TITLE
Adding EF Core Interceptors when using Aspire Client Integration Packages

### DISCUSSION THREAD

**Discussion by mehmetozkaya** [microsoft/aspire#6967] [Q&A] [answered]
I’ve recently added PostgreSQL hosting integration into my AppHost project and after that included the Aspire.Npgsql.EntityFrameworkCore.PostgreSQL client integration NuGet package in my API project. 
While configuring AddNpgsqlDbContext with configureDbContextOptions, I noticed that the serviceProvider parameter isn’t available, which prevents me from resolving dependencies like ISaveChangesInterceptor implementations from the DI container.
In my current setup with traditional DI registration for EF Core, I rely on the serviceProvider parameter to resolve interceptors dynamically, specifically for **DispatchDomainEventsInterceptor**. This interceptor depends on IMediator (from MediatR) to publish domain events, making the serviceProvider crucial for resolving these dependencies.

Here’s a comparison of the two setups:

### Traditional EF Core DI Registration with Interceptors
```
services.AddScoped<ISaveChangesInterceptor, AuditableEntityInterceptor>();
services.AddScoped<ISaveChangesInterceptor, DispatchDomainEventsInterceptor>();

services.AddDbContext<CatalogDbContext>((sp, options) =>
{
    options.AddInterceptors(sp.GetServices<ISaveChangesInterceptor>());
    options.UseNpgsql(connectionString);
});
```

This works perfectly as sp.GetServices<ISaveChangesInterceptor>() dynamically resolves my interceptors, including DispatchDomainEventsInterceptor, which depends on MediatR.
Full code example: [GitHub Link](https://github.com/mehmetozkaya/EshopModularMonoliths/blob/main/src/Modules/Catalog/Catalog/CatalogModule.cs)

### Using Aspire.Npgsql.EntityFrameworkCore.PostgreSQL with AddNpgsqlDbContext
```
builder.Services.AddScoped<ISaveChangesInterceptor, AuditableEntityInterceptor>();
builder.Services.AddScoped<ISaveChangesInterceptor, DispatchDomainEventsInterceptor>();

builder.AddNpgsqlDbContext<CatalogDbContext>(
    "catalogdb",
    configureDbContextOptions: dbContextOptions =>
    {
        dbContextOptions.AddInterceptors([new AuditableEntityInterceptor()]);
        // Issue: Cannot resolve ISaveChangesInterceptor via serviceProvider 
        // DispatchDomainEventsInterceptor requires MediatR injection
    }
);
```
As you can see, without the serviceProvider parameter, I’m unable to dynamically resolve ISaveChangesInterceptor or inject dependencies like IMediator into my interceptors.

### Questions and Suggestions

1. How can I configure EF Core interceptors like DispatchDomainEventsInterceptor without the serviceProvider parameter in this setup? Is there an alternative way to dispatch domain events effectively using the Aspire integration?
2. Are there any plans to include serviceProvider accessibility in the configureDbContextOptions for AddNpgsqlDbContext? This would align better with existing DI patterns and facilitate dynamic resolution of dependencies.
3. If there’s a workaround to accomplish this without the serviceProvider, I’d greatly appreciate the guidance.

I came across a related discussion here: [GitHub Issue #4202](https://github.com/dotnet/aspire/discussions/4202), but it hasn’t been resolved yet. Any insights or updates on this topic would be immensely helpful.

Thank you for your time and effort!

**Comment by davidfowl** [comment]
The answer is here:

https://github.com/dotnet/docs-aspire/issues/1601#issuecomment-2333380717

**Reply by mehmetozkaya** [reply]
Thanks for your reply.
If we go back to the traditional approach with Enrich method, should we define the connection string in appsettings.json again? Or is there a way to use the connectionName defined in the AppHost project?

**Traditional Approach with Connection String and Enrich method:**
```
builder.Services.AddDbContext<CatalogDbContext>((sp, options) =>
{
    options.AddInterceptors(sp.GetServices<ISaveChangesInterceptor>());
    options.UseNpgsql(connectionString); // Should we add the connection string?
});
builder.EnrichNpgsqlDbContext<CatalogDbContext>(); // added for aspire but how do aspire service discovery ?
```
But my main goal is to use Aspire client packages to eliminate manual configurations and benefit from the service discovery feature.

**Looking at the documentation:**
[Add Npgsql Database Context with Enrichment](https://learn.microsoft.com/en-us/dotnet/aspire/database/postgresql-entity-framework-integration?tabs=dotnet-cli#add-npgsql-database-context-with-enrichment)
```
builder.EnrichNpgsqlDbContext<CatalogDbContext>(
    connectionName: "postgresdb",
    configureSettings: settings =>
    {
        settings.DisableRetry = false;
        settings.CommandTimeout = 30;
    });
```
It suggests we can use **connectionName** in the **EnrichNpgsqlDbContext** method, but in practice, the actual method **doesn't seem to have a connectionName** parameter:

Reference: [AspireEFPostgreSqlExtensions.EnrichNpgsqlDbContext<TContext> Method](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.aspireefpostgresqlextensions.enrichnpgsqldbcontext?view=dotnet-aspire-8.0)

Is there any way to use both the connectionName and sp for registering interceptors?

**Reply by davidfowl** [reply, marked as answer]
> Or is there a way to use the connectionName defined in the AppHost project?

`connectionName: "x"` translates to `builder.Configuration.GetConnectionString("x")` so it should be trivial to do the same in your code:

```C#
builder.Services.AddDbContext<CatalogDbContext>((sp, options) =>
{
    options.AddInterceptors(sp.GetServices<ISaveChangesInterceptor>());
    options.UseNpgsql(builder.Configuration.GetConnectionString("x")); // Should we add the connection string?
});
builder.EnrichNpgsqlDbContext<CatalogDbContext>();
```

I see the confusion here, our documentation is wrong https://learn.microsoft.com/en-us/dotnet/aspire/database/postgresql-entity-framework-integration?tabs=dotnet-cli#add-npgsql-database-context-with-enrichment.

cc @IEvangelist

**Reply by mehmetozkaya** [reply]
Thanks so much for the response, now it's very clear from my side. The code below is working. 
```
builder.Services.AddDbContext<CatalogDbContext>((sp, options) =>
{
    options.AddInterceptors(sp.GetServices<ISaveChangesInterceptor>());
    options.UseNpgsql(builder.Configuration.GetConnectionString("catalogdb"));
});
builder.EnrichNpgsqlDbContext<CatalogDbContext>();
```
